sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala

```diff
@@ -187,19 +194,95 @@ object DottyPlugin extends AutoPlugin {
+      // either change sbt's bootclasspath handling or wait until the
+      // dotty-library jar and scala-library jars are merged into one jar.
+      // Meanwhile, let's just put nothing at all on the compiler
+      // bootclasspath, and instead put scala-library and dotty-library on the
+      // compiler classpath. This means that user code could shadow symbols
+      // from these jars but we can live with that for now.
+      classpathOptions := {
+        val old = classpathOptions.value
+        if (isDotty.value)
+          old
+            .withAutoBoot(false)      // we don't put the library on the compiler bootclasspath (as explained above)
+            .withFilterLibrary(false) // ...instead, we put it on the compiler classpath
+        else
+          old
+      },
+      // ... but when running under Java 8, we still need a compiler bootclasspath
+      // that contains the JVM bootclasspath, otherwise sbt incremental
+      // compilation breaks.
+      scalacOptions ++= {
+        if (isDotty.value && !isJavaAtLeast("9"))
+          Seq("-bootclasspath", sys.props("sun.boot.class.path"))
+        else
+          Seq()
+      },
+      managedScalaInstance := {
+        val old = managedScalaInstance.value
+        if (isDotty.value)
+          false
+        else
+          old
+      },
       scalaInstance := Def.taskDyn {
-        val si = scalaInstance.value
-        if (isDotty.value) {
-          Def.task {
-            val dottydocArtifacts = fetchArtifactsOf("dotty-doc", CrossVersion.binary).value
-            val includeArtifact = (f: File) => f.getName.endsWith(".jar")
-            val dottydocJars = dottydocArtifacts.filter(includeArtifact).toArray
-            val allJars = (si.allJars ++ dottydocJars).distinct
-            val loader = new URLClassLoader(Path.toURLs(dottydocJars), si.loader)
-            new ScalaInstance(si.version, loader, si.loaderLibraryOnly, si.libraryJar, si.compilerJar, allJars, si.explicitActual)
-          }
-        } else {
-          Def.task { si }
+        if (isDotty.value) Def.task {
+          val updateReport =
+            fetchArtifactsOf(
+              dependencyResolution.value,
+              scalaModuleInfo.value,
+              updateConfiguration.value,
+              (unresolvedWarningConfiguration in update).value,
+              streams.value.log,
+              scalaOrganization.value %% "dotty-doc" % scalaVersion.value)
+          val scalaLibraryJar = getJar(updateReport,
+            "org.scala-lang", "scala-library", revision = AllPassFilter)
+          val dottyLibraryJar = getJar(updateReport,
+            scalaOrganization.value, s"dotty-library_${scalaBinaryVersion.value}", scalaVersion.value)
+          val compilerJar = getJar(updateReport,
+            scalaOrganization.value, s"dotty-compiler_${scalaBinaryVersion.value}", scalaVersion.value)
+          val allJars =
+            getJars(updateReport, AllPassFilter, AllPassFilter, AllPassFilter)
```

Comment:
The above suggests that `compilerJar` is fetched from the same "bag" as `allJars`. Therefore I don't understand how `allJars` is exempted from the compiler jar, and hence how the NOTE of point 3. in the big comment above is applied. Could you elaborate?

Reply:
`allJars` does contain `compilerJar`, all the jars defined here are used by zinc to form the JVM classpath. The NOTE in 3 talks about scala-library, but I just realized it's not entirely true: I thought that sbt would move the scala-library you depend on from the compiler classpath to the compiler bootclasspath, but actually it removes the scala-library you depend on from the compiler classpath and add to the compiler bootclasspath the scala-library corresponding to your current scalaVersion. This only works because scala guarantees forward and backward bin compat. I'll update the comment to clarify.